### PR TITLE
dist: tools: add headerguard check script

### PR DESCRIPTION
**Pull request by kaspar030**
This PR adds a python script checking for header guards.

Features:
- it outputs diffs that can be directly applied.
- it fixes / adds the commend at the closing ```#endif```
- it adds the file path to the guard, up to ```include/```, or none if there's no ```include``` in the path. (e.g., 'sys/include/a/b/c.h' -> 'A_B_C_H'), 'sys/posix/blah/foo.h' -> 'FOO_H'

One commit  adds some missing header guards (changes made manually),
another commit applies the headercheck output to the codebase.

### CODE REVIEW

**Comment by kaspar030** [comment]
- force-pushed last-minute-fix

**Comment by miri64** [comment]
> it adds the file path to the guard, up to include/, or none if there's no include in the path. (e.g., 'sys/include/a/b/c.h' -> 'A_B_C_H'), 'sys/posix/blah/foo.h' -> 'FOO_H'

Why? I think that puts many constraints on the filename. E.g. if I have a module `blafoo`, I want to call the header `sys/include/blafoo.h` and the interna header `sys/blafoo/_blafoo.h`. If your script works as I expect it to be, this wouldn't be possible, since the leading underscore is removed, right?

**Comment by kaspar030** [comment]
> sys/blafoo/_blafoo.h. If your script works as I expect it to be, this wouldn't be possible, since the leading underscore is removed, right?

No, the script doesn't remove underscores that are in the filename. ```_blafoo.h``` would lead to ```_BLAFOO_H``` as header guard.

**Comment by miri64** [comment]
> No, the script doesn't remove underscores that are in the filename. _blafoo.h would lead to _BLAFOO_H as header guard.

I thought we agreed that this is not allowed Oo. @OlegHahm and @gebart IIRC you were the main people behind this coding convention.

**Comment by gebart** [comment]
@miri64 do we have any header files with leading underscores in the file name, or is this a non-issue?

**Comment by miri64** [comment]
> @miri64 do we have any header files with leading underscores in the file name, or is this a non-issue?

A quick `find` told me no, but this script doesn't check, if a *guard* has leading underscores (which was the intent behind #6438 AFAIK). This would re-introduce such guards (which we removed in a quite lengthy process) as soon one decides to add such header files.

**Comment by kaspar030** [comment]
> This would re-introduce such guards

As of now, it wouldn't, because we don't have header filenames with a leading ```_```.
How would you name the guard of such a file?
This PR is also about finding simple rules that can be enforced.

I was actually surprised how many guards were wrong (copy&pasta, file renames), or which had wrong comments at the closing ```#endif```...

**Comment by miri64** [comment]
> How would you name the guard of such a file?
This PR is also about finding simple rules that can be enforced.

This won't solve potential collisions, but how about `_blafoo.h` => `PRIV_BLAFOO_H`?

**Comment by kaspar030** [comment]
> This won't solve potential collisions, but how about _blafoo.h => PRIV_BLAFOO_H?

I'll add that.
Actually there's only be a collision if two headers with the same name are in the include paths, meaning there's already a problem. AFAIK the local directory would have precedence in that case.

**Comment by kaspar030** [comment]
- addressed leading "_" comment
- added logic to return error if there was a broken/missing header guard

**Comment by kaspar030** [comment]
- fixed error exit code logic

File: dist/tools/headerguards/headerguards.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import os, sys
+import difflib
+#from string import maketrans
+from io import BytesIO, TextIOWrapper
+
+_in = "/-."
+_out = "___"
+
+transtab = str.maketrans(_in, _out)
```
Comment (miri64) [inline dist/tools/headerguards/headerguards.py]: I learned a new python thing \o/ (I usually use `re.sub()` for that).

**Comment by miri64** [comment]
(please squash)

**Comment by kaspar030** [comment]
> Since this enforces a new naming rule for header rules, please update also the coding conventions accordingly.

[Done.](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions#header-guards)

**Comment by kaspar030** [comment]
Thanks for reviewing!

**Pull request merged**